lib/fileset/default.nix

```diff
@@ -0,0 +1,580 @@
+            readDir' = fullPath: filteredPath:
+              let
+                r = builtins.readDir filteredPath;
+              in
+                mapAttrs (name: value:
+                  if ! r ? ${name} then
+                    null
+                  else if r.${name} == "directory" then
+                    readDir' (fullPath + ("/" + name)) (filteredPath + ("/" + name))
+                  else
+                    value
+                ) (builtins.readDir fullPath);
+          in
+            if ! isPath path then
+              throw "gitTracked: Not a path"
+            else if ! lib.path.hasPrefix base path then
+              throw "gitTracked: Path ${toString path} is outside of the root ${toString base}"
+            else create (singletonInternal path (readDir' path importedPath));
+
+        /*
+        Limit a file set to only the files with specific extensions.
+
+            withExts extensionList
+
+        is equivalent to
+
+            fileFilter ({ extension, ... }: lib.elem extension extensionList)
+
+        Type:
+          withExts :: [String] -> FileSet
+        */
+        withExts = extensions:
+          matching (file: elem file.ext extensions);
+
+        /*
+        Damn, these are really cool!
+        */
+        symlinks = matching (file: file.type == "symlink");
+        regular = matching (file: file.type == "regular");
+        unknown = matching (file: file.type == "unknown");
+
+        hidden = directoryMatching (lib.strings.hasPrefix ".");
+
+        nixFiles = withExts [ "nix" ];
+
+        named = name: matching (file: file.name == name);
+
+        /*
+        All files satisfying a predicate function
+        Limit a file set to only the files satisfying a predicate function.
+
+        Type:
+          matching :: ({
+            name :: String,
+            subpath :: String,
+            type :: String,
+            ext :: Null || String,
+            ...
+          } -> Bool) -> FileSet
+        */
+        matching = pred:
+          let
+            # Should not be called with files
+            filterTree = path: tree:
+              mapAttrs (name: type:
+                if type == "directory" || lib.isAttrs type
+                then filterTree (path + ("/" + name)) type
+                else
+                  let
+                    matched = builtins.match ".*\\.(.*)" name;
+                    set = {
+                      inherit name type;
+                      subpath = lib.path.removePrefix base (path + ("/" + name));
+                      ext = lib.mapNullable lib.head matched;
+                    };
+                  in
+                    if pred set
+                    then type
+                    else null
+              ) (directoryEntries path tree);
+          in create (filterTree base "directory");
+
+        /*
+        Limit a file set to only the files for which any path component satisfies some predicate.
```

Comment:
Matching any component seems a bit odd. Do you have a motivating use case for this?
Also since this is about components, I have a feel for what the last component means, but what determines the first?

Reply:
This again changed a bit: The type of `directoryMatches` is now `(String -> Bool) -> Path -> FileSet`. It doesn't consume a fileset anymore, so this isn't really a combinator anymore, but rather it creates a new file set. The components are now relative to the given `Path`. It still won't match filenames though, because I figured it's easier to combine the result of this with another one for filenames:

```nix
union
  (directoryMatches (name: name == ".git") ./.)
  (filter (file: file.name == ".git") ./.)
```

Rather than the other way around if you need _only_ directories.

With this design, it's clear that `lib.fileset.filter` is only running on files (though maybe it should be `lib.fileset.fileFilter` again, name bikeshedding not yet done), while `lib.fileset.directoryMatches` is only running on directories. This will make it much easier to reason about than the previous `builtins.filterSource`-type functions.